added "one-sided" alternative for proportion_confint

mcmrc opened the pull request:
- [ ] closes #9239 
- [x] tests added / passed. 
- [x] code/documentation is well formatted.  
- [x] properly formatted commit message. See 
      [NumPy's guide](https://docs.scipy.org/doc/numpy-1.15.1/dev/gitwash/development_workflow.html#writing-the-commit-message). 

<details>
I added the 'side' parameter to proportion_confint to enable the calculation of a one-sided CI.
The parameter has three options: two-sided(default), upper, and lower.
When either upper or lower is selected, ci_low or ci_upp is returned as np.nan.

pep8speaks commented:
Hello @mcmrc! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`statsmodels/stats/proportion.py`](https://github.com/statsmodels/statsmodels/blob/f832536990609d47cd0b171cfe4d89750dbf1123/statsmodels/stats/proportion.py):

> [Line 195:26](https://github.com/statsmodels/statsmodels/blob/f832536990609d47cd0b171cfe4d89750dbf1123/statsmodels/stats/proportion.py#L195): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 200:67](https://github.com/statsmodels/statsmodels/blob/f832536990609d47cd0b171cfe4d89750dbf1123/statsmodels/stats/proportion.py#L200): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 205:27](https://github.com/statsmodels/statsmodels/blob/f832536990609d47cd0b171cfe4d89750dbf1123/statsmodels/stats/proportion.py#L205): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

* In the file [`statsmodels/stats/tests/test_proportion.py`](https://github.com/statsmodels/statsmodels/blob/f832536990609d47cd0b171cfe4d89750dbf1123/statsmodels/stats/tests/test_proportion.py):

> [Line 1009:1](https://github.com/statsmodels/statsmodels/blob/f832536990609d47cd0b171cfe4d89750dbf1123/statsmodels/stats/tests/test_proportion.py#L1009): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1018:74](https://github.com/statsmodels/statsmodels/blob/f832536990609d47cd0b171cfe4d89750dbf1123/statsmodels/stats/tests/test_proportion.py#L1018): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1022:65](https://github.com/statsmodels/statsmodels/blob/f832536990609d47cd0b171cfe4d89750dbf1123/statsmodels/stats/tests/test_proportion.py#L1022): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1030:69](https://github.com/statsmodels/statsmodels/blob/f832536990609d47cd0b171cfe4d89750dbf1123/statsmodels/stats/tests/test_proportion.py#L1030): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1034:71](https://github.com/statsmodels/statsmodels/blob/f832536990609d47cd0b171cfe4d89750dbf1123/statsmodels/stats/tests/test_proportion.py#L1034): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 1086:1](https://github.com/statsmodels/statsmodels/blob/f832536990609d47cd0b171cfe4d89750dbf1123/statsmodels/stats/tests/test_proportion.py#L1086): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1111:1](https://github.com/statsmodels/statsmodels/blob/f832536990609d47cd0b171cfe4d89750dbf1123/statsmodels/stats/tests/test_proportion.py#L1111): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

josef-pkt commented on statsmodels/stats/tests/test_proportion.py:
```diff
@@ -995,3 +995,116 @@ def test_ci_symmetry_array(count, method):
+    assert_equal(ci_upp, 1.0)
+
+    ci_low, ci_upp = proportion_confint(100, 200)
+    assert_almost_equal(ci_low, 0.4307048087825161)
```
where do the comparison numbers come from?

we use now mainly assert_allclose with atol and/or rtol
(usage of assert_almost_equal comes mostly from before numpy had assert_allclose)

mcmrc replied:
These numbers come from the latest version of statsmodels.stats.proportion_confint. Because I did not alter the calculation logic, those numbers must not be changed.
Should I change all assert_almost_equal to assert_allcose?

josef-pkt replied:
usually we add comparison numbers from other packages, mainly from R for stats.
A `#` comment should mention the source, `regression tests` if numbers are from an earlier statsmodels version.

josef-pkt commented:
about binomtest confidence intervals.

scipy stats has now also confint and alternative options (which it did not have when we added those in statsmodels)

AFAICS, the one sided binom test confints are the same as the beta. (although scipy does rootfinding also in the one-sided case)
unit tests for binomtest confidence interval in scipy with reference numbers from R
https://github.com/scipy/scipy/blob/main/scipy/stats/tests/test_morestats.py#L920

notation/definition
we use the "alternative" keyword also for confints, with `larger` or `smaller` for the outside/alternative region

boundary points
the one-sided intervals should have 0 or 1 for the "open" side instead of nan

josef-pkt commented:
Thanks for the PR

mcmrc commented:
Thank you so much for all the comments.
I will do
- calculate reference numbers from R
- change the argument name to `alternative` and options to `larger` and `smaller`
- use 1 or 0 for open-side.

josef-pkt commented:
great

and add the one-sided intervals to binom test method.
Based on a quick check, I get the same results in scipy as using one of the sides of the beta intervals.
(AFAIR, those are exact in the one-sample case)

jeffreys looks fine to me, it could be replace by isf and ppf in the two one-sided cases. but your current code using `interval` just does a little bit of redundant work. 
(scipy distribution `interval` method is "central", equal tail interval, so standard pattern with doubling alpha still holds.

mcmrc commented:
It seems like the outputs from the current version of statsmodels and R 4.3.3 do not match.

statsmodels: proportion_confint(100, 200, method='binom_test') -> (0.42988371451156804, 0.5701162853923397)
R:  binom.test(100, 200, conf.level=1-0.05)$conf.int -> [1] 0.4286584 0.5713416

Do you have any ideas?

josef-pkt commented:
The problem is base R,
they compute the central, equal tail confidence interval even though the hypothesis test is minlike 

```
proportion_confint(100, 200, method='beta')
(0.42865843090683226, 0.5713415690931678)
```

There is an article that points this out and an R package that provides matching confidence interval
after looking for it: it's exactci, Fay, https://cran.r-project.org/web/packages/exactci/exactci.pdf
e.g. #8233

for method binomtest you need a unit test separate from the other methods, because you cannot check that the one-sided interval can be obtained by doubling alpha.

mcmrc commented:
Thank you, Josef. Understood. scipy.stats.binomtest is not based on a distribution with equal tails.
The past issue is very educative. I'll keep on working on it.

josef-pkt commented:
minlike and central/equal-tail are based on the same distribution, loglikelihood
the equal-tail requires prob <= alpha / 2 in each tail
minlike provides the shortest interval (in uni-modal distributions) but only controls overall coverage and does not control coverage in each tail.

In symmetric distributions like normal, t, both methods are the same, but not in asymmetric distributions that we have in proportions and rates (Binomial and Poisson).

(Coverage in approximate methods can still be slightly offcenter because of the approximation, actual tail probabilites are only approximately alpha/2)

mcmrc commented:
I committed new codes.
- calculated reference numbers from R (BinomCI, binom.test and exactci)
- changed the argument name to `alternative` and options to `smaller` and `larger`
- use 0 or 1 for open-side.
- updated binom_test method with alternative="two-sided"
- diverted method from beta to single-sided binom_test
- refactored test file

mcmrc commented:
Sorry, I've forgotten to remove redundant parts. Please review the latest one.

josef-pkt commented on statsmodels/stats/proportion.py:
```diff
@@ -188,25 +193,40 @@ def _check(x: np.ndarray, name: str) -> np.ndarray:
+                    if alternative == 'two-sided':
+                        print(qi, stats.binomtest(count, nobs, p=qi).pvalue - alpha)
+                        return stats.binomtest(count, nobs, p=qi).pvalue - alpha
+                    elif alternative == 'smaller':
```
AFAICS, these will never be used because of the if above.
so we don't need it

mcmrc replied:
I removed it from the latest commit.

josef-pkt commented on statsmodels/stats/proportion.py:
```diff
@@ -188,21 +193,24 @@ def _check(x: np.ndarray, name: str) -> np.ndarray:
+
+
+    if alternative == 'two-sided':
+        if method != "binom_test":
```
Why not for binom_test?

mcmrc replied:
Because scipy.stats.binomtest uses "two-sided" as default for alternative, the original alpha should be used instead of  alpha / 2.0.
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.binomtest.html

josef-pkt commented:
looks good overall
alternative has maybe larger and smaller switched
I need to go over some details later. (They require thinking which would distract me from my own PR)

mcmrc commented:
Thank you so much for your time.
It is really confusing which side is "larger" or "smaller"; sorry for my mistake.
I read this [docs](https://cran.r-project.org/web/packages/exactci/exactci.pdf) and used "smaller" as equivalent as "less."

![image](https://github.com/statsmodels/statsmodels/assets/20140454/239e2b32-7e3b-497f-b629-93e32c14b98c)

josef-pkt commented:
However, in your example:
**alternative** true probability is less than 0.5, but values less than 0.5 are inside the confidence interval, and not in the alternative rejection region.

I will have to think about this later again. I'm also always getting confused.

something like: at the upper limit of the left (lower) interval we reject the null hypothesis in favor of the alternative hypothesis that the proportion is larger than the value with a pvalue = alpha (in continuous case, and weak inequality in discrete case)
But I need to verify this.

josef-pkt commented:
AFAICS, my intuition was correct
for example https://www.itl.nist.gov/div898/handbook/eda/section3/eda352.htm

in our (my definition in statsmodels) the one-sided `alternative` also corresponds to the one-sided `alternative` in the hypothesis test.
If the true value is in/inside the confidence interval, then we reject the null hypothesis in favor of the alternative hypothesis with probability less than alpha in repeated sampling.

This means, the rejection region is in the direction of the alternative.
For the lower tail confidence interval, the rejection region is in the direction of the "larger" alternative.
For the upper tail confidence interval, the rejection region is in the direction of the "smaller"or "less" alternative.

mcmrc commented:
I see that. So, the word 'alternative' corresponds to the reject region.

josef-pkt commented:
Can you squash this?  earlier commits are not relevant

final version looks good to me and can be merged

Thanks for the PR

mcmrc commented:
I've been always helped by statsmodels in my research.
Thank you so much for your time and effort, Josef.

josef-pkt commented:
unrelated unit test failure, merging

josef-pkt commented:
If you are interested, then you could add alternative in the same way to poisson rates
https://www.statsmodels.org/dev/generated/statsmodels.stats.rates.confint_poisson.html

Those are all `central`, so unit test could be to just compare with the two-sided interval.

mcmrc commented:
Let me take that.
I created a new issue and will make a PR.
#9254
